sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/statefulOperators.scala

```diff
@@ -642,9 +647,7 @@ case class SessionWindowStateStoreSaveExec(
         // Assumption: watermark predicates must be non-empty if append mode is allowed
         case Some(Append) =>
           allUpdatesTimeMs += timeTakenMs {
-            val filteredIter = applyRemovingRowsOlderThanWatermark(iter,
```

Comment:
Hmm, I think rows older than watermark are dropped at earlier operators? This looks like a no-op?

Reply:
Oh yea, actually in the below iterator, we remove the values by watermark as condition. It is same as here.

Reply:
The difference is that here we only filtered out sessions which new events are appeared for the grouping keys, whereas below we filtered out sessions from all state rows.

To trigger the problem, the grouping key should have existing sessions which has the end time earlier than the watermark. It's a bit tricky based of how watermark works in SS and how no-data batch is triggered. These sections are expected to be evicted on no-data batch so tricky to reproduce.